### ISSUE TITLE
Set up a job to check O*NET frameworks for updates periodically

### ISSUE THREAD

**Issue by woodkri** [cassproject/cass-editor#646]
So frameworks don't have to be manually checked and re-imported.

**Comment by Lomilar** [comment]
The approach to this currently is to set up a node JS script tied to a job that does the import.

**Comment by woodkri** [comment]
Capturing:
I tried just a plain import via node and the 4 smaller frameworks had no problems. The technology skills gave me 2 different errors depending on which server I tried, both in the post to selectedServer/ctdlasn.
dev.api returned this right away:
```
<html>
<head><title>413 Request Entity Too Large</title></head>
<body bgcolor="white">
<center><h1>413 Request Entity Too Large</h1></center>
<hr><center>nginx/1.14.0 (Ubuntu)</center>
</body>
</html>
```
And regular dev returned this after a few minutes:
```
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>502 Proxy Error</title>
</head><body>
<h1>Proxy Error</h1>
<p>The proxy server received an invalid
response from an upstream server.<br />
The proxy server could not handle the request<p>Reason: <strong>Error reading from remote server</strong></p></p>
<hr>
<address>Apache/2.4.41 (Ubuntu) Server at dev.cassproject.org Port 443</address>
</body></html>
```

**Comment by mparsons-ce** [comment]
I have set up an endpoint for testing. 
https://sandbox.credentialengine.org/publisher/api/Framework/RepublishFramework?password=TBD&frameworkCtid=ce-.......

**Password:**
I had set it up to require a password for initial testing (to be supplied). Or I could change to require the API key from the accounts site. 
**FrameworkCtid:** 
the CTID for a framework in CaSS that has been previously published. 

The endpoint will validate that the CTID is for a framework that has been previously published. 
If valid, the framework will be added to a pending queue for publishing later in the day.

**Comment by woodkri** [comment]
I updated the script to check the dateModified and import if the source has a more recent date than the version in CaSS. Still to do here:

- Sort out the timeout issue on staging (see my comments in #643 for details) preventing import of Technology Skills
- On successful import, send to publishing queue (will need the password for this)
- Set up the job to automatically run the script periodically (I have node set up on staging, so all this job needs to do is cd into the "Update ONET script" folder on the desktop and run "node checkForOnetUpdates.js" by command line
- Repeat on production

**Comment by mparsons-ce** [comment]
Actions
- ApiKey: Decided to use an api key in the header instead of a password. I will send the api keys to Kristin separately. 
- Schedule: Kristin will set up a folder and script to initiate the check. 
 -- Michael will set up a scheduled job to call the script
- A scheduled task to run batch publish job which will check the pending republish queue

**Comment by woodkri** [comment]
The same timeout issue is occurring on production for the Technology Skills framework.

**Comment by Lomilar** [comment]
The timeout issue doesn't mean it didn't import, just that it timed out and you can't see the result. Can be remedied by ensuring IIS has a lengthy timeout specified, but would likely affect all service calls, which isn't always the best choice.

**Comment by woodkri** [comment]
I know the import is still happening, but when we're running this by script and it times out, the script never terminates. We also don't know when the imports finish, to be able to add the ids to the republish queue at that point. I suppose we could just make sure the publish queue is checked significantly after this script runs, but we still would have the problem of the command prompt waiting for the server response that never arrives.

**Comment by Lomilar** [comment]
Would it be possible to ignore the import success method and check the framework URL every minute to see if the framework has changed?

Non-traditional, but it's a workaround.

**Comment by woodkri** [comment]
That looks like it should work. @mparsons-ce Is the republish queue endpoint set up on staging and/or production? Testing on staging, I'm getting a message: "No HTTP resource was found that matches the request URI 'https://staging.credentialengine.org/publisher/api/Framework/RepublishFramework?frameworkCtid=ce-07c257d6-9119-11e8-b852-782bcb5df6ac'."

**Comment by mparsons-ce** [comment]
I added the endpoint to staging. 
https://staging.credentialengine.org/publisher/api/Framework/RepublishFramework?password=TBD&frameworkCtid=ce-07c25f74-9119-11e8-b852-782bcb5df6ac

I did not get to updating the endpoint to use an apiKey. 
If you want to do some quick testing I can send you the current 'test' password. 
I just got out of a meeting and will work on adding the use of an apiKey.

**Comment by woodkri** [comment]
Thanks @mparsons-ce ! You should see all 5 frameworks in the republish queue (some may have multiple entries as I was testing the post to the endpoint). I'm not sure if you want to set up a job to run the script on staging or just on production, but the script is ready on staging. It can be run by cd-ing into the "Update ONET script" folder on the desktop and running "node checkForOnetUpdates.js" by command line.

Let me know when the endpoint is ready on production and I'll update the script there as well.

**Comment by mparsons-ce** [comment]
@woodkri 
The cassAdmin desktop folder is protected, so I created the folder: C:\CredentialEngine\CaSS_Scripts and copied **Update ONET script**, on both staging and production.
C:\CredentialEngine\CaSS_Scripts\Update ONET script

I added a scheduled task to invoke the latter at 5pm central (Mon-Fri for now).
I added a scheduled task to invoke the republish process at 7pm central. 

I checked and there were 5 records in the queue. The 'queuing' process checks for an existing pending item, so will not add duplicates. I ran the republish task and 4 out of 5 were successful. Any guesses on the failed task?

For production, I added a scheduled task to invoke the CaSS script at 5pm central (Daily).

**Comment by woodkri** [comment]
@mparsons-ce Is the endpoint ready on production? I haven't updated the script there yet with the technology skills workaround and the republish post but I can go ahead and do so if so. The scheduled task shouldn't be run on production until that's been updated.

**Comment by mparsons-ce** [comment]
@woodkri Yes, the endpoint is ready on production.

FYI: I found the error on attempt to export the Technology Skills framework:
TypeError: Cannot get property "length" of null

If I entered the following URL in browser while logging into the staging VM, I get the latter error. 
https://staging.credentialengine.org/api/ceasn/5b3e205efac3276dd7edfdd1345d8530

However, when I use the staging publisher (https://staging.credentialengine.org/publisher/Competencies), select and then export the framework I don't get an error?

Just in case I went back to the staging VM, and again tried to navigate to the export URL and this time it worked!

**Comment by woodkri** [comment]
The script has been updated on production. You should now see the 5 frameworks in the publishing queue.

**Comment by mparsons-ce** [comment]
@woodkri I noticed that all (well I checked the first 100 or so) of the dateCreated for Technology Skills have dates with a day of 2020-08-20. I checked a few of the other frameworks, and they had a variety of dateCreated. 
Does the dateCreated get modified on import to CaSS or is it the date from O*Net?

**Comment by woodkri** [comment]
If there is no dateCreated provided, we use the current date/time during the import process. I believe they provided a dateCreated for competencies in the other frameworks but not in Technology Skills.

**Comment by mparsons-ce** [comment]
@woodkri On staging, the CaSS job is scheduled for 5pm central. The job is set to stop if it runs for more than two hours:
![image](https://user-images.githubusercontent.com/58825866/91058953-b92e0700-e5ee-11ea-87d2-3c6851a0bfe6.png)

Is there no 'internal' stop mechanism? Or due to the issues with Technology Skills, does the job have to continue just in case?
If the latter is true, is it reasonable to stop the job after two hours? Or can it be stopped sooner?
The question would of course also apply to production.

**Comment by woodkri** [comment]
I took a look at the task scheduler since the script should be terminating now that I've added the workaround from above. I think the issue is that the Action is specified as starting a program at the script location. It needs to be run with the command `node checkForOnetUpdates.js` in a console window; just starting up the file isn't going to run the script. If it's not possible to program in that command to the Task Scheduler action, one of us can write up a batch file that runs the command. I haven't worked with the task scheduler before to know what you can and can't do there.

**Comment by mparsons-ce** [comment]
OK, I added a cmd file to start the process and reference the latter in the scheduler
![image](https://user-images.githubusercontent.com/58825866/91067155-2a25ec80-e5f8-11ea-95ca-0ea0ff1856b6.png)

I tried running the cmd file and saw messages that four frameworks were up to date, and that one was queued for republishing.

**Comment by woodkri** [comment]
Great!

**Comment by mparsons-ce** [comment]
@woodkri We have been checking the dateModified for the O*Net frameworks. Four of the five have an internal date modified from April, and one, Worker Characteristics - Abilities has a date of Nov. 13, 2019. On the O*Net site, the latter has a date modified of April 14, 2020. 
How does the internal dateModified of a framework get set? Does it come from an internal property of the O*Net document?

Related, if these frameworks have not been updated since April, how come there have been a few new additions to the queue yesterday?

**Comment by woodkri** [comment]
I'm trying to figure out what's happening with the Abilities framework. If I try to get the framework data the usual ways, I see the date from 2019. If I go directly to Elasticsearch though (localhost:9200/...) I see the date from April 2020.

These dates come from the O*Net data. To my knowledge, the frameworks haven't been updated since April with the exception of the date modified which was just added. Because the date modified is what we're checking, these registered as changed frameworks, which triggered the import and republish. If they end up in the queue again after they've been updated and republished, let me know and I will investigate.

**Comment by mparsons-ce** [comment]
Verified as working properly in production.

**Comment by mparsons-ce** [comment]
How to add a new O*Net framework:

- Log into Production web server
- Change to directory: **C:\CredentialEngine\CaSS_Scripts\Update ONET script**
- Edit: **checkForOnetUpdates.js**
- Add to **var urls**:
- Current content
var urls = ["https://www.onetcenter.org/ctdlasn/graph/ce-07c257d6-9119-11e8-b852-782bcb5df6ac", "https://www.onetcenter.org/ctdlasn/graph/ce-07c25f74-9119-11e8-b852-782bcb5df6ac", "https://www.onetcenter.org/ctdlasn/graph/ce-07c264d7-9119-11e8-b852-782bcb5df6ac", "https://www.onetcenter.org/ctdlasn/graph/ce-07c27a0f-9119-11e8-b852-782bcb5df6ac", "https://www.onetcenter.org/ctdlasn/graph/ce-9fab4187-d8e7-11e9-8250-782bcb5df6ac"];

**Comment by mparsons-ce** [comment]
@woodkri 
I have noticed that in the last week the Cross-Functional skills, Knowledge, and technology skills frameworks were scheduled for republishing. 
We had the usual problems with technology skills. 
I checked the other two and there was **no change to the modified date**. 
On publish to the registry, the registry process checks if the current payload is the same as the current payload. If they are the same, no update is done. Both of the latter were marked as **not changed**. 
Technically there is no harm is attempting to republish a framework that has not changed.
However, I thought you should be notified in case there are some adjustments that could be made to the script that checks for O*Net changes.

**Comment by woodkri** [comment]
I see why that was happening for Technology Skills and made a small edit for that one. I have no idea why the other two would have triggered the republish. I updated the cassproject npm package in case that makes a difference. Let me know if this keeps happening and I'll take another look.

**Comment by mparsons-ce** [comment]
@woodkri 
This is still regularly happening. Last night:
O*NET Content Model Worker Requirements - Knowledge: U.S. Department of Labor (DOL) (no change)

**Comment by woodkri** [comment]
Ok I think I figured out what's happening. I had written the script so that if a framework hadn't yet been imported, we could add it to the list and import for the first time this way. Every once in a while the framework get was failing, so it appeared the framework didn't exist and the import triggered. I have removed that section of code, and we'll just need to make sure we import any future frameworks for the first time through the interface before relying on the script to import them.
